src/metabase/query_processor/streaming/csv.clj

```diff
@@ -72,50 +73,72 @@
+                                       (qp.pivot.postprocess/add-totals-settings viz-settings)
+                                       qp.pivot.postprocess/add-pivot-measures))
               ;; col-names are created later when exporting a pivot table, so only create them if there are no pivot options
-              col-names (when-not opts (common/column-titles ordered-cols (::mb.viz/column-settings viz-settings) format-rows?))]
-          ;; when pivot options exist, we want to save them to access later when processing the complete set of results for export.
-          (when opts
-            (reset! pivot-options (merge {:pivot-rows []
-                                          :pivot-cols []} opts)))
+              col-names          (when-not opts (common/column-titles ordered-cols (::mb.viz/column-settings viz-settings) format-rows?))
+              pivot-grouping-key (qp.pivot.postprocess/pivot-grouping-key col-names)]
+
+          ;; initialize the pivot-data
+          ;; If exporting pivoted, init the pivot data structure
+          ;; Otherwise, just store the pivot-grouping key index
+          (when (and pivot? pivot-export-options)
+            (reset! pivot-data (qp.pivot.postprocess/init-pivot opts)))
+          (when pivot-grouping-key
+            (swap! pivot-data assoc :pivot-grouping pivot-grouping-key))
+
           (vreset! ordered-formatters
                    (if format-rows?
                      (mapv #(formatter/create-formatter results_timezone % viz-settings) ordered-cols)
                      (vec (repeat (count ordered-cols) identity))))
+
           ;; write the column names for non-pivot tables
           (when col-names
-            (write-csv writer [col-names])
-            (.flush writer))))
+            (let [header (m/remove-nth (or pivot-grouping-key (inc (count col-names))) col-names)]
+              (write-csv writer [header])
+              (.flush writer)))))
 
       (write-row! [_ row _row-num _ {:keys [output-order]}]
         (let [ordered-row (if output-order
                             (let [row-v (into [] row)]
-                              (for [i output-order] (row-v i)))
-                            row)
-              xf-row      (perf/mapv (fn [formatter r]
-                                       (formatter (common/format-value r)))
-                                     @ordered-formatters ordered-row)]
-          (if @pivot-options
-            ;; if we're processing a pivot result, we don't write it out yet, just store it
-            ;; so that we can post process the full set of results in finish!
-            (swap! rows! conj xf-row)
-            (do
-              (write-csv writer [xf-row])
-              (.flush writer)))))
+                              (into [] (for [i output-order] (row-v i))))
+                            row)]
+          (if (contains? @pivot-data :config)
+            ;; if we're processing a pivot result, we don't write it out yet, just aggregate it
```

Comment:
My previous attempt at pivot post-processing stored every row in an atom and did the post-process aggregations/building of the table in one go, that proved to be memory-inefficient.

This approach still stores data in an atom, but it will only grow as big as the final table size anyway, and each row is just aggregated one at a time (so effectively only totals + distinct values are stored).

Of course, the nature of pivot tables is combinatorial and depends on the number of unique values in each pivot row and pivot column, so a user can still construct a table that is quite large, but this is a better approach than before.